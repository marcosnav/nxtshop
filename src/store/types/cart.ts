import { Commit } from 'vuex';

export namespace Cart {
  export interface State {
    products: string[];
  }

  /**
   * Payloads
   */
  export interface UpdatePayload {
    item: string;
    operation: 'ADD' | 'REMOVE';
  }

  /**
   * Mutations
   */
  type MutationPayload = UpdatePayload;

  export enum MutationType {
    UPDATE_CART = 'UPDATE_CART',
  }

  export function commit(
    c: Commit,
    type: Cart.MutationType.UPDATE_CART,
    p: Cart.UpdatePayload
  ): void;

  export function commit(
    c: Commit,
    type: Cart.MutationType,
    payload: MutationPayload
  ) {
    c(type, payload);
  }

  /**
   * Actions
   */

  export enum ActionType {
    ADD_TO_CART = 'ADD_TO_CART',
  }
}
