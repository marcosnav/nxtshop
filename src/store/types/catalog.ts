export namespace Catalog {

  export interface ProductMeta {
    id: string;
    name: string;
    tagline: string;
    thumbnail: string;
  }
  export interface Product extends ProductMeta {
    description: string;
    options: ProductOption[];
  }

  export interface ProductQuickView extends ProductMeta {
    price: string;
  }

  export interface ProductOption {
    id: string;
    description: string;
    price: number;
  }
}
