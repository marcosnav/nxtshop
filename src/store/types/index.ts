export * from './root';
export * from './cart';
export * from './catalog';
