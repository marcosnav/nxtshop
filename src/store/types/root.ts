export namespace Root {
  export interface State {
    version: string;
  }
}
