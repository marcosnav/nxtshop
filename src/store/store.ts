import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import { CartModule } from './modules';
import { Root } from './types';

Vue.use(Vuex);

const store: StoreOptions<Root.State> = {
  state: {
    version: '0.1.0',
  },
  modules: {
    cart: CartModule,
  },
};

export default new Vuex.Store<Root.State>(store);
