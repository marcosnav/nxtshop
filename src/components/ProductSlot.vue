<template>
  <div class="ProductSlot">
    <product-thumb :src="product.thumbnail" :alt="product.name" />
    <section class="ProductSlot__meta">
      <price-tag>${{ product.price }}</price-tag>
      <section-title>
        {{ product.name }}
      </section-title>
      <tagline>
        {{ product.tagline }}
      </tagline>
    </section>
    <section class="ProductSlot__actions">
      <one-button small textline>
        Add to Wishlist
      </one-button>
      <one-button small>
        Get It Now
      </one-button>
    </section>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
import ProductThumb from './ProductThumb.vue';
import SectionTitle from './SectionTitle.vue';
import Tagline from './Tagline.vue';
import PriceTag from './PriceTag.vue';
import OneButton from './OneButton.vue';
import { Catalog } from '@/store/types';

export default Vue.extend({
  components: {
    ProductThumb,
    SectionTitle,
    Tagline,
    PriceTag,
    OneButton,
  },

  props: {
    product: Object as () => Catalog.ProductQuickView,
  },
});
</script>

<style lang="sass" scoped>
.ProductSlot
  background: var(--subtle-gray)
  border-radius: 6px
  box-sizing: border-box
  padding: 16px
  position: relative
  width: 100%

  &__meta
    margin-bottom: 26px
    margin-top: -47px

  &__actions
    display: flex
    justify-content: center

    & > *
      margin: 5px 0
</style>
