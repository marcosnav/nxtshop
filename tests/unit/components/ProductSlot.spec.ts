import { mount } from '@vue/test-utils';
import ProductSlot from '@/components/ProductSlot.vue';
import { Catalog } from '@/store/types';

const productMockData: Catalog.ProductQuickView = {
  id: '1234-1234-1234',
  name: 'SmartWatch X',
  tagline: 'Smart tech wrapping your wrist',
  thumbnail: 'https://bla.bla.image.com/tech.jpg',
  price: '599',
};

describe('<ProductSlot>', () => {
  const component = mount(ProductSlot, {
    propsData: {
      product: productMockData,
    },
  });

  test('renders properly', () => {
    const priceTag = component.find('.PriceTag');
    const sectionTitle = component.find('.SectionTitle');
    const tagline = component.find('.Tagline');
    expect(component.classes('ProductSlot')).toBeTruthy();
    expect(priceTag.text()).toBe('$599');
    expect(sectionTitle.text()).toBe(productMockData.name);
    expect(tagline.text()).toBe(productMockData.tagline);
    expect(component.element).toMatchSnapshot();
  });
});
