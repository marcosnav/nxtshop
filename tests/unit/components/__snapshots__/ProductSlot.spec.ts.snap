// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductSlot> renders properly 1`] = `
<div
  class="ProductSlot"
>
  <figure
    class="ProductThumb"
  >
    <img
      alt="SmartWatch X"
      class="ProductThumb__image"
      src="https://bla.bla.image.com/tech.jpg"
    />
  </figure>
   
  <section
    class="ProductSlot__meta"
  >
    <div
      class="PriceTag"
    >
      $599
    </div>
     
    <h1
      class="SectionTitle"
    >
      
      SmartWatch X
    
    </h1>
     
    <h2
      class="Tagline"
    >
      
      Smart tech wrapping your wrist
    
    </h2>
  </section>
   
  <section
    class="ProductSlot__actions"
  >
    <button
      class="OneButton OneButton--small OneButton--textline"
    >
      
      Add to Wishlist
    
    </button>
     
    <button
      class="OneButton OneButton--small"
    >
      
      Get It Now
    
    </button>
  </section>
</div>
`;
